Throw an error when side-effectful logic happens in a React component tree

trueadm opened the pull request:
Release notes: the React reconciler now throw a FatalError upon encountering side-effects in a render

This PR revamps the current React system's restrictions for what you can and can't do during the React reconcilation phase. This is a pretty large update but provides much better boundaries between what is "safe" and not "safe", thus reducing the constraints.

1. A new error `ReconcilerRenderBailOut` is thrown if something occurs in the React reconciliation that causes the render to fail and it's a hard fail – no recovering and continuing.
2. If you mutate a binding/object outside the render phase, given the React component render phase is meant to be "pure", a `ReconcilerRenderBailOut` will be thrown.
3. If you `throw` during the React reconciliation phase, again a `ReconcilerRenderBailOut` will be thrown.

In the future, we should maybe loosen the constraints around all this and maybe allow `throw`, but right now it's causing too much friction. We should attempt to make React components render phase as pure as possible – as it results in much better optimizations by a compiler because we can assert far more without things tripping us up.

Another point, regarding (1), is that we should ideally be able to recover from the error thrown in Prepack. The reason we can't and is something that might be a very deep issue in Prepack, is that effects don't properly restore when we have nested PossiblyNormalCompletions at work. Bindings get mutated on records from changes made within `evaluateForEffects` but never get reset when in nested PossiblyNormalCompletion. If I remove all the things that can cause `PossiblyNormalCompletion`s then everything works fine and bindings do get restored. We can remove the constraint on (1) once we've found and fixed that issue.

trueadm commented:
@NTillmann if you check out this PR and try this in `fb-www/input.js`:

```js
require("react");

__evaluatePureFunction(function() {
  var React = require("react");

  var lazy = null;

  function createLazy() {
    if (!lazy) {
      lazy = {};
    }
  }

  function URI(arg) {
    if (!arg) {
      return;
    }
    var obj;
    if (arg === 0) {
      obj = {};
    }
    if (!{}[obj]) {
      throw new Error("no");
    }
    this.foo = obj;
  }

  function Middle(props) {
    var foo = props.href.foo;
    URI(foo);
    var href = URI(foo);
    createLazy();
    return href;
  }

  function App(props) {
    return React.createElement(Middle, {
      href: new URI("http://foo.com/" + props.x)
    });
  }

  __optimizeReactComponentTree(App, {
    firstRenderOnly: true
  });

  module.exports = App;
});
```

I get an issue where it still tries to visit an `AbstractValue` that was created in the React component render's effects. The thing is, these effects are never applied and should be reverted because an impure component tree was detected. For some reason, the binding value was not properly restored to `null`.

trueadm commented on src/react/reconcilation.js:
```diff
@@ -198,6 +199,9 @@ export class Reconciler {
       )
     );
+    if (!checkIfRenderWasPure(effects, evaluatedRootNode)) {
+      return null;
```
If the render was impure, we skip applying it's effects. So the effects are never used and should be restored.

trueadm commented on src/react/utils.js:
```diff
@@ -850,3 +850,71 @@ export function sanitizeReactElementForFirstRenderOnly(realm: Realm, reactElemen
+function checkForWriteBindings(effects: Effects): void {
+  let [result, , writeBindings] = effects.data;
+  for (let [binding] of writeBindings) {
+    debugger;
```
Intentionally left in so you can see it correctly finds an impure binding modification.

trueadm commented on src/react/reconcilation.js:
```diff
@@ -198,6 +199,16 @@ export class Reconciler {
+        // This should not be abstract, it should be null
+        // if you put a return null at the top of this function
+        // this will not be the same
+        invariant(false, ":(");
```
This is to help debugging.

hermanventer commented:
Perhaps rename __sideEffect to __safeSideEffect to emphasize the responsibility of the author to ensure that nothing bad will happen despite the side effect.

NTillmann commented on scripts/test-react.js:
```diff
@@ -81,6 +81,7 @@ function runTestSuite(outputJsx, shouldTranspileSource) {
     reactOutput: outputJsx ? "jsx" : "create-element",
     inlineExpressions: true,
     invariantLevel: 0,
+    simpleClosures: true,
```
This will unlock a whole new set of bugs.

gaearon replied:
We *do* use it for `debug-fb-www` script though. So this moves us closer to how we currently generate the bundles.

Does it mean it's too unstable and we shouldn't?

sebmarkbage replied:
It is completely broken. Let's remove it and fix the normal stuff instead. E.g. by optimizing it if we need to. We need to do something because it'll definitely generate wrong code in current scenarios.

trueadm commented:
@hermanventer Fixed the cycle length.

hermanventer commented on src/react/reconcilation.js:
```diff
@@ -160,29 +200,22 @@ export class Reconciler {
-          evaluatedRootNode.status = "BAIL-OUT";
+          message = `Failed to optimize React component tree for "${
+            evaluatedRootNode.name
+          }" due to an expected bail-out: ${error.message}`;
```
Would the "expected bail-out" language reach users of the React compiler? If so, they may not know what to make of it.

hermanventer commented on src/react/reconcilation.js:
```diff
@@ -1448,4 +1475,10 @@ export class Reconciler {
   }
+
+  _throwIfRenderWasNotPure(effects: Effects, evaluatedNode: ReactEvaluatedNode): void {
+    // TODO: should we also check realm.savedEffects?
```
Hopefully effects lists everything the render did. It would be weird and brittle if additional effects were still lurking in realm.savedEffects.

trueadm commented:
@NTillmann I looked at `__residual` and it doesn't do what we want. I originally applied the havoc logic to `__residual`, but then thought it might be risky as these are used on RN in production which might be dangerous right now. We can align what we mean and what to do in a design meeting – there should be some form on convergence though as I'd imagine you'd want to use `safeSideEffect` in IR.

trueadm commented on src/intrinsics/prepack/global.js:
```diff
@@ -400,4 +401,31 @@ export default function(realm: Realm): void {
+        Havoc.value(realm, func.$BoundTargetFunction);
+      }
+      // given we don't actually want the modified bindings/properties from the havoc
+      // as this is a side-effect and such, we can discard the modified bindings/properties
```
I updated all the comments around this – just in case anyone was wondering.

NTillmann commented on src/intrinsics/prepack/global.js:
```diff
@@ -400,4 +401,38 @@ export default function(realm: Realm): void {
+      }
+      // given we don't actually want the modified bindings/properties from the havoc
+      // as this is a side-effect and such, we can discard the modified bindings/properties
+      realm.modifiedBindings = originalModifiedBindings;
```
I don't fully understand the intention here, but this looks wrong and dangerous to me. So `Havoc` changed some bindings, but removing the new entries from `realm.modifiedBindings` means that these changed bindings won't participate in proper tracking and undoing of effects later on.

trueadm replied:
We could diff the changes to bindings and then ignore them specifically when checking for side-effects? I originally tried that but ran into some strange errors.

sebmarkbage replied:
Yea, this is way too fragile.

The indention here is that we want to havoc something outside the pure scope which we never do in any other case, and we want that havoc to be permanent so it shouldn't be reverted.

I'm not sure this is the best strategy though. Not sure havocing outside the pure function actually works as expected. Will need to think about that.

sebmarkbage commented on src/utils/havoc.js:
```diff
@@ -307,7 +309,7 @@ class ObjectValueHavocingVisitor {
       this.visitDeclarativeEnvironmentRecordBinding(record, remainingHavocedBindings);
 
-      if (record instanceof FunctionEnvironmentRecord) {
+      if (record instanceof FunctionEnvironmentRecord && !sideEffectfulWrapper) {
```
So this part is because we want to havoc everything that is captured? That will only effect bindings but not objects that are captured.

So `__safeSideEffect(() => x++)` might work but not `__safeSideEffect(() => obj.x++` since only obj will be havoced, not what's inside it. But if you do it deeply, it'll pull in the whole environment with it including the global scope.

It's also weird that order matters now. Any optimized functions before this will read this value assuming initial value and any optimized functions that runs after this will assume havoced value.

sebmarkbage commented:
We should take a look if we really need this. I'd prefer that the fix is that we move both the binding/object that we're reading from and the write to be abstract.

If not, I think we should have to wrap the binding/object itself in something that makes it unsafe, not the mutation. Reading a mutable value is as bad as writing to it.

So in code like this:

```js
var x = 0;
function Foo() {
  x++;
}
```

Instead of turning the mutation into safe, how about instead, we havoc it in the module scope? It's never safe to read from it and mutations needs to be tracked.

```js
var x = 0;
__havoc(() => { x; });
function Foo() {
  x++;
}
```

Currently havocing fatals in a non pure scope but this special function could by-pass that.

This would turn the `x++` into an assignment statement in the generator instead of a modified binding:

https://github.com/facebook/prepack/blob/master/src/environment.js#L295

Which will by-pass your check.

Not all side-effects are covered by modified bindings and properties unfortunately. A lot of them comes from generator entries such as this one. That's another issue to consider.

trueadm commented:
The alternative to address @sebmarkbage and @NTillmann concerns for this PR might be to removed `__safeSideEffect` entirely. We'd need to refactor all the tests though, but maybe that's a good thing. What do you think @gaearon?

gaearon commented:
I think that in practice we should be able to exclude (or fix or transform) product code that currently relies on mutations, so we should start by just forbidding them, and fix our tests if we need to. Then if we find a situation in product code where a directive like this really buys us something we can consider adding support for this back.

trueadm commented:
Let's do that then. I'm happy removing `__safeSideEffect` – I never liked it, I just thought everyone else would want it and I'm somewhat happy that everyone seems to be equally supportive of not having it. Plus it adds additional complexity that we would be best to avoid right now. If anyone else would like to apply the changes to this PR to remove it, feel free!

gaearon commented:
Should we start working on getting non-simpleClosures mode working with our tests before landing this?

trueadm commented:
We might not need to. The tests in this PR needed `simpleClosures` because of `__safeSideEffect` but if we're removing it we might not need to change that flag – as long as we fix up all the failing tests.

sebmarkbage commented:
@gaearon That sounds fine. I think a feature that havocs in module scope before we optimize like my proposal above would actually work fine and isn't too bad of a feature. There might be cases with closures that are currently broken due to serializeBinding issues but those will also be broken with any other havocing of closures.

Btw, does this check cover things like Map?

```js
var x = new Map();
function Foo() {
  x.set('x', 'y');
}
```
Maybe add a test?

trueadm commented:
@sebmarkbage @gaearon I've removed `__safeSideEffect` from this PR. I've also updated tests that now fail because of side-effects in the render phase and rather than removing them (I did remove one firstRender test as it was actually a bad test), I added a test assertion called `expectReconcilerRenderBailOut` that validates the error threw. The error is validated and the message it has is snapshotted so can detect regressions where the error message changes.

trueadm commented:
Furthermore, we should follow up with @sebmarkbage's proposal in another PR. I'm not sure the best way of automatically doing this – but maybe we can be explicit like in the examples @sebmarkbage gave. Either way, it's not in the scope of this PR and we should design around that in a follow up.

gaearon commented on src/react/reconcilation.js:
```diff
@@ -160,29 +207,22 @@ export class Reconciler {
+          message = `Failed to optimize React component tree for "${
+            evaluatedRootNode.name
+          }" due to an exception thrown during render phase`;
         } else {
```
Seems like when we have a Prepack bug (not even an invariant, but a JS error like accessing a property on undefined), we get here. This doesn't seem great because we lose the original call stack. For example I got a message like this:

```
Error: Failed to optimize React component tree for "FeedStoryUFISummary" due to evaluation error: Cannot read property '$Get' of undefined
    at renderComponentTree (/Users/gaearon/p/prepack/src/react/reconcilation.js:225:15)
    at Realm.evaluateForEffects (/Users/gaearon/p/prepack/src/realm.js:761:15)
```

Could we preserve the stack, or better, only "wrap" errors that are known to be intentional (FatalError?), and aren't just bugs in Prepack?

trueadm replied:
Good point, I'll do that for sure.

gaearon commented on src/react/reconcilation.js:
```diff
@@ -96,6 +103,47 @@ export type ComponentTreeState = {
+        `Failed to optimize React component tree for "${
+          evaluatedNode.name
+        }" due to an exception thrown during render phase`,
+        evaluatedNode
```
When I get this in a bundle I'm testing, there doesn't seem to be no indication of where the error happened.

The component name seems to point to the top-level component which isn't very useful. Instead, I want to know where `throw` is located, and its JS stack.

trueadm replied:
We can probably work out from the binding, what function it was in. I'm not sure how we can make a call stack again though as this happens after all the evaluation has occured.

gaearon commented:
This looks like a mutation in render phase to me.

Why does it compile without error?

```js
require("react");

__evaluatePureFunction(function() {
  var React = require("react");

  var foo = {};

  function App(props) {
    if (foo.hasOwnProperty(props.x)) {
      foo[props.x] = 0;
    }
    const x = foo[props.x]++;
    return x;
  }

  __optimizeReactComponentTree(App);

  module.exports = App;
});
```

gaearon commented:
As discussed offline, we'll need a more precise callsite reporting for `throw` because it's hard to hunt them down.

sebmarkbage commented:
There is already a place where this happens. 
https://github.com/facebook/prepack/blob/master/src/realm.js#L1301
https://github.com/facebook/prepack/blob/master/src/realm.js#L1329

Just have to add one that checks for isInPureScope and if the object/closure is part of the newly created objects in `createdObjectsTrackedForLeaks` and fatal if they're not.

trueadm commented:
@sebmarkbage I tried that originally but it didn’t work for me. :( this why I had to check the effects after the entire render

sebmarkbage commented:
It might be because we catch FatalError in a bunch of scenarios which lets us bail out. Maybe we need a EvenMoreFatalError.

Alternatively, we can also just log it to diagnostics without fataling.

gaearon commented on src/react/reconcilation.js:
```diff
@@ -163,47 +165,27 @@ export class Reconciler {
+        () =>
+          // TODO: (sebmarkbage): You could use the return value of this to detect if there are any mutations on objects other
+          // than newly created ones. Then log those to the error logger. That'll help us track violations in
+          // components. :)
```
Is it time to address this? :-)

trueadm replied:
Hehe, this PR means we can finally remove this TODO!

trueadm commented:
@sebmarkbage are you happy with this PR now?

gaearon commented:
I'm going to give it a test now

gaearon commented on src/react/reconcilation.js:
```diff
@@ -1476,4 +1482,26 @@ export class Reconciler {
+      if (typeof key === "string") {
+        name = `"${key}"`;
+      }
+      throw new UnsupportedSideEffect(`side-effects from mutating the object ${name}${location}`);
```
"from mutating a property" maybe?

I'm seeing an error like

>Failed to render React component "Tooltip" due to side-effects from mutating the object "(\d)(\d{3})($|\D)"

when it points to code like

```js
var _regexCache = {};
```

I assume the `name` is the property name here, not the object name.

gaearon commented on src/values/ObjectValue.js:
```diff
@@ -150,7 +150,7 @@ export default class ObjectValue extends ConcreteValue {
           // whether this.$IsClassPrototype has been initialized as a proxy for
           // object initialization in general.
-          invariant(this.$IsClassPrototype === undefined || !this.isHavocedObject(), "cannot mutate a havoced object");
+          // invariant(this.$IsClassPrototype === undefined || !this.isHavocedObject(), "cannot mutate a havoced object");
```
Don't forget to revert this

sebmarkbage commented on src/realm.js:
```diff
@@ -1324,6 +1372,19 @@ export class Realm {
+      this.isInPureScope() &&
+      this.reportSideEffectCallback !== undefined &&
+      this.createdObjectsTrackedForLeaks !== undefined &&
+      binding.object instanceof ObjectValue &&
```
Can this ever be AbstractObjectValue here? If not, this should be an invariant. If we can, then we need to handle it because it can cover up mutations if it is wrapped by an abstract value.

trueadm replied:
Will add those checks

sebmarkbage commented on src/realm.js:
```diff
@@ -1295,16 +1310,49 @@ export class Realm {
+        let env = binding.environment;
+
+        if (
+          !(env instanceof FunctionEnvironmentRecord) ||
```
The binding can be other environment records that are not safe. Like with, block scoped variables (let/const)...

trueadm replied:
Will add other bindings, do you know which ones you were thinking of?

sebmarkbage replied:
I think you can add an invariant here that `env instanceof DeclarativeEnvironmentRecord`. If it's one of the other environments, like the ones that write to objects (`with` and global), it should not get recorded here. It'll be recorded as a modified property. That's probably a good comment to add.

However, it might not be a function environment record.

If it is a let/const, it should be a DeclarativeEnvironmentRecord. To figure out what function it belongs to you have to traverse its parent until you hit a function or global. That's the function that was used to create the environment record. That function is what you use to look up in the map.

sebmarkbage commented on src/realm.js:
```diff
@@ -1295,16 +1310,49 @@ export class Realm {
-    if (this.modifiedBindings !== undefined && !this.modifiedBindings.has(binding))
+
+    const isEnvSafe = root => {
+      let context = this.getRunningContext();
```
I don't think the currently running context matters. What matters is where the binding was defined. You need to pass in the environment from the binding here.

I'll try to think of a test case but pretty sure this is wrong.

trueadm replied:
A test case would be awesome :)

sebmarkbage replied:
```js
var x = 0;
function foo() {
  x++; // not safe
}
function bar() {
  foo();
}
__pure(bar);
```

```js
var x = 0;
function foo() {
  function foo2() {
    x++; // not safe
  }
  foo2();
}
function bar() {
  foo();
}
__pure(bar);
```

```js
function foo() {
  var x = 0;
  function foo2() {
    x++; // safe
  }
  foo2();
}
function bar() {
  foo();
}
__pure(bar);
```

```js
function foo() {
  var x = 0;
  function foo2() {
    function foo3() {
      x++; // safe
    }
    foo3();
  }
  foo2();
}
function bar() {
  foo();
}
__pure(bar);
```

trueadm replied:
All these tests pass with the current code.

trueadm commented on src/realm.js:
```diff
@@ -316,6 +318,9 @@ export class Realm {
   createdObjects: void | CreatedObjects;
   createdObjectsTrackedForLeaks: void | CreatedObjects;
   reportObjectGetOwnProperties: void | (ObjectValue => void);
+  reportSideEffectCallback:
```
I'll handle these in a follow up.
